### DISCUSSION TITLE
when will the condition set for a plugin be fired?

### DISCUSSION THREAD

**Discussion by jdhao** [wbthomason/packer.nvim#534] [Q&A] [answered]
I am using the following config to load vim-fugitive conditionaly:

```lua
 use {'tpope/vim-fugitive', cond =
 function()
   local res = vim.fn.system('git rev-parse --is-inside-work-tree')
   if string.find(res, 'true') then
     return true
   else
     return false
   end
 end}
```
It only works when I first open nvim, i.e., fugitive will be loaded if the directory is a Git directory, otherwise fugitive will not be loaded.

It seems that the `cond` is only run when we firt enter Nvim? How can I make it also run when I change working directory inside Nvim?

**Comment by wbthomason** [comment, marked as answer]
Ah - yes, `cond` is only run when the compiled file is sourced (so, when you enter Neovim). The easiest way to do what you want is to load on the `DirChanged` event using the `event` key.

**Reply by jdhao** [reply]
If I use `DirChanged` event, I can not check if I am inside a Git repository (fugitive will be blindly loaded whenever I change current directory). Can I combine `DirChanged` with condition? I think what I want is as follows:

+ If we open nvim inside a Git repository, then load fugitive.
+ If we open nvim in a non-git repo and change the current working directory to a git repo, then load fugitive.

Can we achieve this with packer? I thought `event` and `cond` are exclusive, is it?

**Reply by wbthomason** [reply]
Fair point. You can achieve this, but not by simply using the `event` and `cond` keys.
- You can make an autocommand *separate* from `packer` that triggers on `DirChanged` and runs your "am I in a Git repository" function. Your function can then trigger a `User` autocommand, e.g. `User InGitRepo`.
- You can then load `fugitive` on `event = User InGitRepo`.

We could conceivably also make `cond` be checked on every attempted load of a plugin (i.e. it would run not just at startup). This wouldn't be hard to do, but I'd want to think about whether this would break other use cases. I think it wouldn't (one could always adapt their `cond` function to check for the original condition as well as a "have we already started", etc.). However, right now, this isn't how things work.

**Reply by jdhao** [reply]
Thanks for your kind mentor! Following your instruction, I have implemented the idea.

```vim
augroup git_repo_check
  autocmd!
  autocmd VimEnter,DirChanged * call utils#Inside_git_repo() 
augroup END

" inside ~/.config/nvim/autoload/utils.vim
" Check if we are inside a Git repo
function! utils#Inside_git_repo() abort
  let res = system('git rev-parse --is-inside-work-tree')
  if match(res, 'true') == -1
    return v:false
  else
    " Trigger a speical user autocmd
    doautocmd User InGitRepo
    return v:true
  endif
endfunction
```

The setting for packer:

```lua
use {'tpope/vim-fugitive', event = 'User InGitRepo'}
```

**Reply by WieeRd** [reply]
To anyone who stumbled upon this page while seeking for optimization,  
The code above by @jdhao is now possible in pure Lua.
```lua
-- requires Neovim v0.7+
local vim = vim

local function check_git_repo()
  local cmd = "git rev-parse --is-inside-work-tree"
  if vim.fn.system(cmd) == "true\n" then
    vim.api.nvim_exec_autocmds("User", { pattern = "InGitRepo" })
    return true  -- removes autocmd after lazy loading git related plugins
  end
end

vim.api.nvim_create_autocmd(
  { "VimEnter", "DirChanged" },
  { callback = function() vim.schedule(check_git_repo) end }
)
```
`vim.schedule` is required if you want to benefit from lazy-loading and improve startup time.  
Calling `system()` actually takes 6 times longer than loading `vim-fugitive`!  
(at least it did on my PC, taking about 6ms and 1ms each)
